components/camera/videosource/webcam.go

```diff
@@ -193,54 +194,141 @@ func makeConstraints(attrs *WebcamAttrs, debug bool, logger golog.Logger) mediad
+	if err != nil {
+		return true, errors.Wrap(err, "cannot get properties from media source")
+	}
+	// github.com/pion/mediadevices connects to the OS to get the props for a driver. On disconnect props will be empty.
+	return len(props) != 0, nil
+
+}
 
-	goutils.PanicCapturingGo(func() {
-		src, err := camera.SourceFromCamera(cam)
+// reconfigureCamera creates a new camera and reconfigures the given reconfigurable camera.
+func reconfigureCamera(
+	ctx context.Context,
+	oldCam resource.Reconfigurable,
+	attrs *WebcamAttrs,
+	label string,
+	logger golog.Logger,
+) (err error) {
+	newCam, err := findCamera(ctx, attrs, label, logger)
+	defer func() {
 		if err != nil {
-			logger.Debugw("cannot get source from camera", "error", err)
-			return
+			goutils.UncheckedError(goutils.TryClose(ctx, newCam))
 		}
+	}()
+	if err != nil {
+		return errors.Wrap(err, "cannot make camera")
+	}
 
-		defer func() {
-			logger.Debug("closing camera")
-			goutils.UncheckedError(cam.Close(ctx))
-		}()
+	if err = oldCam.Reconfigure(ctx, newCam); err != nil {
+		return errors.Wrap(err, "cannot reconfigure camera")
+	}
+	return
+}
+
+// NewWebcamSource returns a new source based on a webcam discovered from the given attributes.
+func NewWebcamSource(ctx context.Context, attrs *WebcamAttrs, logger golog.Logger) (camera.Camera, error) {
+	cam, err := findCamera(ctx, attrs, attrs.Path, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot find video source for camera")
+	}
+
+	label := attrs.Path
+	if label == "" {
+		label = getLabelFromCamera(utils.UnwrapProxy(cam).(camera.Camera), logger)
+	}
+
+	const wait = 500 * time.Millisecond
+	webcamConnectionFn := func() {
+		defer goutils.UncheckedError(goutils.TryClose(ctx, cam))
 		for {
-			if goutils.SelectContextOrWait(ctx, 500*time.Millisecond) {
-				// mediadevices connects to the OS to get the properties for a driver. If the OS no longer knows
-				// about a specific driver then properties will be empty, and we can safely assume the driver no
-				// longer exists and should be closed.
-				if len(gostream.PropertiesFromMediaSource[image.Image, prop.Video](src)) == 0 {
-					return
-				}
-			} else {
+			if !goutils.SelectContextOrWait(ctx, wait) {
 				return
 			}
+
+			ok, err := isConnected(cam)
+			if err != nil {
+				logger.Debugw("cannot determine camera status", "error", err)
+				continue
+			}
+
+			if !ok {
+				goutils.UncheckedError(goutils.TryClose(ctx, cam))
+				logger.Debugw("camera disconnected", "label", label)
+				for {
+					if !goutils.SelectContextOrWait(ctx, wait) {
+						return
+					}
+					if err := reconfigureCamera(ctx, cam, attrs, label, logger); err != nil {
+						logger.Debugw("cannot reconfigure camera", "label", label, "error", err)
+						continue
+					}
+					logger.Debugw("camera connected", "label", label)
+					break
+				}
+			}
 		}
-	})
+	}
 
-	return cam, nil
+	goutils.ManagedGo(webcamConnectionFn, webcamConnectionFn) // restart on panics
```

Comment:
the 2nd parameter is not for restarts, it's for what's called when the goroutine finishes gracefully (no panics). Instead this should be used to either close a channel or decrement a waitgroup so that you can cleanly close out a webcam, which is currently missing.

Reply:
Removed the restart on panic logic. The camera was closed on 313 but that has been moved to inside of `reconfigureCamera`. It's also closed in the defer statement on the first line of `webcamConnectionFn`.